core/templates/services/connection-service.service.ts

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service to check for the internet connection.
+ */
+
+import { EventEmitter, Injectable } from '@angular/core';
+import { Subscription, timer } from 'rxjs';
+import { delay, retryWhen, switchMap, tap } from 'rxjs/operators';
+// eslint-disable-next-line oppia/disallow-httpclient
+import { HttpClient } from '@angular/common/http';
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+/**
+ * Instance of this interface is used to report current connection status.
+ */
+export interface ConnectionState {
+  /**
+   * "True" if browser has network connection. Determined by Window
+   * objects "online" / "offline" events.
+   */
+  hasNetworkConnection: boolean;
+  /**
+   * "True" if browser has Internet access. Determined by heartbeat system
+   * which periodically makes request to heartbeat Url.
+   */
+  hasInternetAccess: boolean;
+}
+
+export interface ConnectionCheckResponse {
+  isInternetConnected: boolean;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ConnectionService {
+  private INTERNET_CONNECTIVITY_CHECK_INTERVAL_MILLISECS: number = 3500;
+  private MAX_MILLISECS_TO_WAIT_UNTIL_NEXT_CONNECTIVITY_CHECK: number = 7000;
+  private checkConnectionUrl: string = '/connectivity/check';
+
+  private _stateChangeEventEmitter = new EventEmitter<ConnectionState>();
```

Comment:
"stateChange" part can be a bit confusing in the context of an exploration. Can we change this to be more specific about what "state" in this context means please?

Reply:
Changed it to ```_connectionStateChangeEventEmitter```.